PythonCodeConventions.md

```diff
@@ -1,25 +1,33 @@
 Python Code Conventions
 =======
 
-PEP8 is the code style guide for Python code used by the main python distribution.
-This style guide can be found at:
-https://www.python.org/dev/peps/pep-0008/
+General
+---
+Ultimaker adheres to the PEP8 coding guidelines as described here: https://www.python.org/dev/peps/pep-0008/,
 
-Our coding conventions follow the PEP style guide, with the exceptions stated in the generic code_conventions.md provided in this repository. This also means that for rules regarding class naming, documentation, etc., the rules described in the general code convention document MUST be used.
+** Important: **
+Also check the code_conventions.md file for the generic rules, regardless of language.
 
-For convenience, these are re-listed here. In case of conflict, the generic code_conventions are considered to be leading.
-* Always use spaces. While PEP8 allows tabs for backwards compatibility, we do not.
-* Strings are double quotes. While Python allows single and double quoted strings, and PEP8 only says "pick a rule and stick to it". We decided to do double quotes to match C++.
-* "Maximum line length" - PEP8 specifies a maximum line length of 79 characters. At Ultimaker we have no official line length limit, although we strive to keep lines below 120 characters.
-* Naming convention:
-  * Function names in PEP8 are in lower_case_with_underscores. At Ultimaker we decided to match our C++ coding convention and use mixedCase (CapitalizedWords by initial lowercase character).
-  * Parameters, variables and members follow the lower_case_with_underscores convention. PEP8 has no recommendation on this.
+With the following adjustments, exceptions and specific choices:
+* See the 'line breaking' section below for specific suggested PEP8 style choices where function calls and definitions exceed the soft line length limit of 120 characters.
+* MUST apply the following string formatting:
+  * Multiple elements:
+``` python
+"{foo}_{bar}".format(foo = "foo", bar = "bar")
+```
+ * Single element:
+``` python
+"{}_bar".format("foo")
+```
+* SHOULD avoid circular dependencies as much as possible. Remove when encountered according to boy-scout rules.
```

Comment:
Avoiding circular dependencies is a generic good coding practice. I don't understand why it is mentioned at all and when it would be important to mention it should be in the code_conventions.md document and not here as an exception from the generic rule.

Reply:
This was discussed and suggested within a Python context only. I sure can see this as a general rule, but I'm not familiar with other languages covered by the generic guidelines to be able to judge if this holds up for all affected languages. @ChrisTerBeke can you?

Reply:
I think this counts for every language. At least it also does for JavaScript/TypeScript (the only other language next to Python we use in Cloud and Connect).

Reply:
@ChrisTerBeke I believe this collides with forward declarations in Cpp? I'd like to stick to Python rules and not open up another can of worms a.t.m.